modules/client-common/src/main/java/org/apache/ignite/client/proto/ClientMessageDecoder.java

```diff
@@ -78,13 +73,10 @@ private boolean readMagic(ByteBuf byteBuf) {
         if (byteBuf.readableBytes() < MAGIC_BYTES.length)
             return false;
 
-        assert data.length == MAGIC_BYTES.length;
-
-        byteBuf.readBytes(data, 0, MAGIC_BYTES.length);
+        var data = new byte[MAGIC_BYTES.length];
```

Comment:
Seems that logic with `magicFailed` and `magicDecoded` works only one time. And seemd that it is not needed at all
Just add 
```
 if (byteBuff.readableBytes() < MAGIC_BYTES.length) {
            return false
 }
```

Reply:
1. The first 4 bytes of every client connection must contain the magic header, as per [IEP-76](https://cwiki.apache.org/confluence/display/IGNITE/IEP-76+Thin+Client+Protocol+for+Ignite+3.0)

2. New `ClientMessageDecoder` instance is created for every `Channel` (connection) thanks to `ChannelInitializer`